### PR TITLE
refactor(agent)!: Use Pydantic models for `Agent` process output

### PR DESCRIPTION
**Pull request by Pwuts**
## **User description**
Make prompting & parsing more straightforward and improve typing throughout the application by using pydantic models for the agent's cycle output.

This is partially the result of work on XML support; the rest is to clean up connected code.

### Changes 🏗️

* Introduce `BaseAgentActionProposal`, `OneShotAgentActionProposal`, and `AssistantThoughts` models to replace `ThoughtProcessResponse`, `DEFAULT_RESPONSE_SCHEMA`
* Refactor and clean up code because now we don't need to do as much type checking everywhere
* Tweak `OneShot` response format instruction

Granular:

* `autogpt.agents.prompt_strategies.one_shot`
  * Replace ThoughtProcessResponse`, `DEFAULT_RESPONSE_SCHEMA` and parsing logic by `AssistantThoughts` and `OneShotAgentActionProposal`
  * (TANGENTIAL) Move response format instruction into main system prompt message
  * (TANGENTIAL) Adjust response format instruction

* `autogpt.agents.base`
  * Add `BaseAgentActionProposal` base model -> replace `ThoughtProcessOutput`
  * Change signature of `execute` method to accept `BaseAgentActionProposal` instead of separate `command_name` and `command_args`
  * Add `do_not_execute(proposal, feedback)` abstract method, replacing `execute("human_feedback", ..., feedback)`

* Move `history` definition from `BaseAgentSettings` to `AgentSettings` (the only place where it's used anyway)

* `autogpt.models`
  * Add `.utils` > `ModelWithSummary` base model
  * Make the models in `.action_history` (`Episode`, `EpisodicActionHistory`) generic with a type parameter for the type of `Episode.action`

* `autogpt.core.resource.model_providers.schema`
  * Add `__str__` to `AssistantFunctionCall` which pretty-prints the function call

All other changes are a direct result of the changes above.

## BREAKING CHANGE:
* Due to the change in `autogpt.models.action_history`, the application after this change will be unable to load/resume agents from before this change and vice versa.
* The `additional_output` field in the response of `execute_step` has changed slightly:
  * Type of `.thoughts.plan` has changed from `str` to `list[str]`
  * `.command` -> `.use_tool`
  * `.command.args` -> `.use_tool.arguments`

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [x] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

___

## **Type**
enhancement

___

## **Description**
- Introduced new Pydantic models `BaseAgentActionProposal`, `AssistantThoughts`, and `OneShotAgentActionProposal` to standardize agent actions and thoughts.
- Refactored agent methods to utilize these new models, enhancing the clarity and maintainability of the code.
- Updated the agent's prompt strategies to align with new models, improving the parsing and validation of agent responses.
- Adjusted related components and utilities to support these changes, ensuring compatibility across the agent's operation.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Update Agent Module Imports and Exports</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
autogpts/autogpt/autogpt/agents/__init__.py

<li>Updated imports to include new models and removed old ones.<br> <li> Expanded <code>__all__</code> to export new action proposal models.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7116/files#diff-f4d114ef0e9064b4e3c6a9a560b89eed267ea412e41310dc8028679f73d3adbb">+8/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>agent.py</strong><dd><code>Refactor Agent Methods to Use New Action Proposal Model</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
autogpts/autogpt/autogpt/agents/agent.py

<li>Refactored methods to use new <code>OneShotAgentActionProposal</code> model.<br> <li> Updated method signatures and logic to accommodate new model <br>structures.<br> <li> Removed old command execution logic and replaced with new tool <br>execution.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7116/files#diff-9224bfb0fac9d6e9ef7e4533bacf7dee4f89ba02339088c95f5e7cfc3455dd94">+68/-67</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>base.py</strong><dd><code>Introduce BaseAgentActionProposal and Update Base Agent</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
autogpts/autogpt/autogpt/agents/base.py

<li>Introduced <code>BaseAgentActionProposal</code> replacing old <code>ThoughtProcessOutput</code>.<br> <li> Updated abstract methods to use new action proposal model.<br> <li> Removed old command-related types.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7116/files#diff-a0a9fb0388413a4926ab73838636d31fcd967cd62744c05820dfb387a2c9f3f9">+17/-19</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>watchdog.py</strong><dd><code>Update Watchdog Component to Use New Action Proposal Model</code></dd></summary>
<hr>
      
autogpts/autogpt/autogpt/agents/features/watchdog.py

- Updated method signatures to use the new action proposal model.

</details>
    

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7116/files#diff-6037b84435aa76e0bde54bb588a0411e3b174a4aec7a17c1499b04754017a647">+11/-11</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>one_shot.py</strong><dd><code>Define New Models and Update One Shot Prompt Strategy</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
autogpts/autogpt/autogpt/agents/prompt_strategies/one_shot.py

<li>Defined new models <code>AssistantThoughts</code> and <code>OneShotAgentActionProposal</code>.<br> <li> Updated prompt strategy to use new response schema and parsing logic.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7116/files#diff-c8c3d9b541659d0906f0e03a0b6ed211f5aa3f243393556a9d711ec4a908d193">+45/-172</a></td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: autogpts/autogpt/autogpt/models/utils.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from abc import ABC, abstractmethod
+
+from pydantic import BaseModel
+
+
+class ModelWithSummary(BaseModel, ABC):
+    @abstractmethod
+    def summary(self) -> str:
+        """Should produce a human readable summary of the model content."""
+        pass
```
Comment (kcze) [inline autogpts/autogpt/autogpt/models/utils.py:10]: Unimportant; no need for `pass` when there's a docstring + no newline at the end

**Review by kcze — APPROVED** [review]
Looks good, I like the simplification of code this brings by removing `extract_command` and `DEFAULT_RESPONSE_SCHEMA`.
Unrelated: this Codecov `was not covered by tests` every few lines in some files is quite distracting 😄

**Comment by Pwuts** [comment]
> Unrelated: this Codecov was not covered by tests every few lines in some files is quite distracting 😄

@kcze press `a` in the diff view and they disappear :)

File: autogpts/autogpt/autogpt/app/agent_protocol_server.py [hunk 2]
```diff
@@ -215,38 +213,37 @@
...
-            execute_command = agent.event_history.current_episode.action.name
-            execute_command_args = agent.event_history.current_episode.action.args
+            last_proposal = agent.event_history.current_episode.action
             execute_approved = not user_input
 
             logger.debug(
-                f"Agent proposed command"
-                f" {execute_command}({fmt_kwargs(execute_command_args)})."
+                f"Agent proposed command {last_proposal.use_tool})."
                 f" User input/feedback: {repr(user_input)}"
```
[the remark this answers is not in the document]
Comment (Pwuts) [reply autogpts/autogpt/autogpt/app/agent_protocol_server.py]: This suggestion is bullshit, the closing parenthesis is there but shouldn't be.

**Pull request merged**